packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Appending the wp-block-heading to before rendering the stored `core/heading` block contents.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading block content.
+ *
+ * For example, the following block content:
+ *  <h2 class="align-left">Hello World</h2>
+ *
+ * Would be transformed to:
+ *  <h2 class="align-left wp-block-heading">Hello World</h2>
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+	
+	$p = new WP_HTML_Tag_Processor( $content );
+
+	$header_tags = array( 'H1', 'H2', 'H3', 'H4', 'H5', 'H6' );
+	while ( $p->next_tag() ) {
+		if ( in_array( $p->get_tag(), $header_tags, true ) ) {
+			$p->add_class( 'wp-block-heading' );
```

Comment:
```suggestion
			$p->add_class( 'wp-block-heading' );
			break;
```

Reply:
so I believe that originally, as written, the function modified all heading tags, something which I tried to reproduce in my suggested code. `preg_replace_callback` will continue to scan unless given a `$limit` parameter.

glad this is clearer now what the behavior should be.